### PR TITLE
Feature/colored coin

### PR DESCRIPTION
**Pull request by Yamaguchi**
This pull request introduces the colored coin feature to electrs-tapyrus.

2 APIs are updated 
- `blockchain.scrpthash.listunspent` returns a list of all colored and uncolored utxos.
- `blockchain.scrpthash.get_balance`  returns a list of balances for each ColorIdentifier.

And new 2 APIs are added.
- `blockchain.scrpthash.listcoloredunspent` returns a list of colored utxos.
- `blockchain.scrpthash.listuncoloredunspent` returns a list of uncolored (i.e native TPC) utxos.

APIs for Open Assets Protocol is enabled.
I think it is desired to provide the function to switch this Open Assets Protocol.
But this pull request has not contained it yet.

### CODE REVIEW

File: src/rpc.rs [hunk 1]
```diff
@@ -267,6 +288,20 @@ impl Connection {
         Ok(unspent_from_status(&self.query.status(&script_hash[..])?))
     }
 
+    fn blockchain_scripthash_listcoloredunspent(&self, params: &[Value]) -> Result<Value> {
```
Comment (rantan) [inline src/rpc.rs:291]: Does this RPC support `color_id` argument that is described in protocol.md? 

I could not find the place where gets `params.get(1)`.

File: src/index.rs [hunk 2]
```diff
@@ -81,11 +82,19 @@ pub struct TxOutRow {
...
 
 impl TxOutRow {
-    pub fn new(txid: &Txid, output: &TxOut) -> TxOutRow {
+    pub fn new(txid: &Txid, output: &TxOut, colored: bool) -> TxOutRow {
+        let script = if colored {
+            let (_color_id, script) = split_colored_script(&output.script_pubkey)
+                .expect("Expected colored script(cp2pkh or cp2sh) but script is not colored");
+            script
+        } else {
+            Script::from(Script::from(Vec::from(&output.script_pubkey[..])))
```
[the remark this answers is not in the document]
Comment (Yamaguchi) [reply src/index.rs]: Fixed at bce5dfcff3af8bcbf981825f0415689dd1f1cbea

File: src/rpc.rs (hunk 1 above)
Comment (Yamaguchi) [reply src/rpc.rs:304]: Fixed at 166f8e45ae826b179ce8cf320d5676d294017af8

**Comment by rantan** [comment]
I have a question about the usage of the listcoloredunspent RPC.

My understanding of the usage is that a scripthash that is a first argument of the RPC should not include OP_COLOR and color id at the head of the script. Is this correct?

If that is the case, I think users can get specific colored UTXOs by offering a scripthash that includes OP_COLOR and the color id in the script.
In this case, the query to scan DB is able to get specific colored UTXOs at first without filtering. So I guess this way is profit the performance.
If this idea is correct, I think the color id argument of listcoloredunspent is unnecessary. (then, I think we need to describe the usage to the document.)

I am sorry for such a mention after you worked for the color id support.

**Comment by Yamaguchi** [comment]
> My understanding of the usage is that a scripthash that is a first argument of the RPC should not include OP_COLOR and color id at the head of the script. Is this correct?

Scripthash is usually hash of p2pkh or p2sh, which does not include OP_COLOR and color_id, but the client may include OP_COLOR and color id. 
 
> users can get specific colored UTXOs by offering a scripthash that includes OP_COLOR and the color id in the script.
> In this case, the query to scan DB is able to get specific colored UTXOs at first without filtering
Yes, but spliting script(p2pkh, p2sh) and color id is useful because the application 

> So I guess this way is profit the performance.
Right. I can fix the implementation so that performance is not lost by specifying Color.

**Comment by rantan** [comment]
Thanks, I understand the benefit of the color id argument.

Could you add a description for the scripthash about the below things to the document?
* In general, listcoloredunspent receive p2pkh, p2sh script if a user wants to get color UTXOs.
* By using cp2pkh, cp2sh script hash, user can get specific colored coin's UTXOs with a bit good performance.

**Comment by Yamaguchi** [comment]
> Could you add a description for the scripthash about the below things to the document?
> 
> * In general, listcoloredunspent receives p2pkh, p2sh script, even if a user wants to get color UTXOs.
> * By using cp2pkh, cp2sh script hash, user can get specific colored coin's UTXOs with a bit good performance.

Add descripton for it.
828c00666577b3f629090503ab9bc0266c25eb19

**Pull request merged**